src/main/java/com/puppycrawl/tools/checkstyle/site/XdocsTemplateParser.java

```diff
@@ -0,0 +1,203 @@
+public class XdocsTemplateParser extends XdocParser {
+
+    /** The source content of the input reader. Used to pass into macros. */
+    private String sourceContent;
+
+    /** A macro name. */
+    private String macroName;
+
+    /** The macro parameters. */
+    private Map<String, Object> macroParameters = new HashMap<>();
+
+    @Override
+    public void parse(Reader source, Sink sink, String reference) throws ParseException {
+        sourceContent = null;
+
+        try (Reader reader = source) {
+            final StringWriter contentWriter = new StringWriter();
+            IOUtil.copy(reader, contentWriter);
+            sourceContent = contentWriter.toString();
+            super.parse(new StringReader(sourceContent), sink, reference);
+        }
+        catch (IOException ioException) {
+            throw new ParseException("Error reading the input source", ioException);
+        }
+        finally {
+            sourceContent = null;
+        }
+    }
+
+    @Override
+    protected void handleStartTag(XmlPullParser parser, Sink sink) throws MacroExecutionException {
+        final String tagName = parser.getName();
+        if (tagName.equals(DOCUMENT_TAG.toString())) {
+            sink.body();
+            sink.rawText(parser.getText());
+        }
+        else if (tagName.equals(MACRO_TAG.toString()) && !isSecondParsing()) {
+            processMacroStart(parser);
+            setIgnorableWhitespace(true);
+        }
+        else if (tagName.equals(PARAM.toString()) && !isSecondParsing()) {
+            processParamStart(parser, sink);
+        }
+        else {
+            sink.rawText(parser.getText());
+        }
+    }
+
+    @Override
+    protected void handleEndTag(XmlPullParser parser, Sink sink) throws MacroExecutionException {
+        final String tagName = parser.getName();
+        if (tagName.equals(DOCUMENT_TAG.toString())) {
+            sink.rawText(parser.getText());
+            sink.body_();
+        }
+        else if (tagName.equals(MACRO_TAG.toString())
+                && macroName != null
+                && !macroName.isEmpty()
+                && !isSecondParsing()) {
+            processMacroEnd(sink);
+            setIgnorableWhitespace(false);
+        }
+        else if (!tagName.equals(PARAM.toString())) {
+            sink.rawText(parser.getText());
+        }
+    }
+
+    /**
+     * Handle the opening tag of a macro. Gather the macro name and parameters.
+     *
+     * @param parser the xml parser.
+     * @throws MacroExecutionException if the macro name is not specified.
+     */
+    private void processMacroStart(XmlPullParser parser) throws MacroExecutionException {
+        macroName = parser.getAttributeValue(null, Attribute.NAME.toString());
+
+        if (macroParameters == null) {
+            macroParameters = new HashMap<>();
+        }
+
+        if (macroName == null || macroName.isEmpty()) {
+            throw new MacroExecutionException(String.format(Locale.ROOT,
```

Comment:
Please make `Strung message`. To allow user clearly see that we throw exception by simply message, no cause.
It becomes matter only if message creation is multiple lines.

Please do for other cases you create in this PR.